language/move-stdlib/nursery/sources/Capability.move

```diff
@@ -0,0 +1,153 @@
+///
+/// ```
+///   struct ProtectedFeature { ... } // this can be just a type tag, or actually some protected data
+///
+///   public fun initialize(s: &signer) {
+///     // Create capability. This happens once at module initialization time.
+///     Capability::create<ProtectedFeature>(s);
+///   }
+///
+///   public fun do_something(s: &signer) {
+///     // Acquire the capability. This is the authorization step. Must have a signer to do so.
+///     let cap = Capability::acquire<ProtectedFeature>(s);
+///     // Pass the capability on to functions which require authorization.
+///     critical(cap);
+///   }
+///
+///   fun critical(cap: Capability::Cap<ProtectedFeature>) {
+///     // Authorization guaranteed by construction -- no verification needed!
+///     ...
+///   }
+/// ```
+///
+/// Notice that a key feature of capabilities is that they do not require extra verification steps
+/// to ensure authorization is valid.
+///
+/// ## Delegation
+///
+/// Capabilities come with the optional feature of *delegation*. Via delegation, an owner of a capability
+/// can designate another signer to be also capable of acquiring the capability. Like the original owner,
+/// the delegate needs to present his signer to obtain the capability in his transactions. Delegation can
+/// be revoked, removing this access right from the delegate.
+///
+/// While the basic authorization mechanism for delegates is the same as with core capabilities, the
+/// target of delegation might be subject of restrictions which need to be specified and verified. This can
+/// be done via global invariants in the specification language. For example, in order to prevent delegation
+/// all together for a capability, one can use the following invariant:
+///
+/// ```
+///   invariant forall a: address where exists<CapState<ProtectedFeature>>(addr):
+///               len(Capability::spec_delegates<ProtectedFeature>(a)) == 0;
+/// ```
+///
+/// Similarly, the following invariant would enforce that delegates, if existent, must satisfy a certain
+/// predicate:
+///
+/// ```
+///   invariant forall a: address where exists<CapState<ProtectedFeature>>(addr):
+///               forall d in Capability::spec_delegates<ProtectedFeature>(a):
+///                  is_valid_delegate_for_protected_feature(d);
+/// ```
+///
+module Std::Capability {
+    use Std::Errors;
+    use Std::Signer;
+    use Std::Vector;
+
+    const ECAP: u64 = 0;
+    const EDELEGATE: u64 = 1;
+
+    /// The token representing an acquired capability. Cannot be stored in memory, but copied and dropped freely.
+    struct Cap<phantom Feature> has copy, drop {
+        root: address
+    }
+
+    /// An internal data structure for representing a configured capability.
+    struct CapState<phantom Feature> has key {
+        delegates: vector<address>
+    }
+
+    /// An internal data structure for representing a configured delegated capability.
+    struct CapDelegateState<phantom Feature> has key {
```

Comment:
It would be great if this extends so that one can have multiple same capabilities from different roots. For example, VASP1 and VASP2 can delegate their KeyRotationCap to the same RecoveryAddress.

Reply:
Noted for later treatment.

Reply:
Perhaps, it will also require the change of the `acquire` API eventually like: `public fun acquire<Feature>(requester: &signer, root: address): Cap<Feature>`. So, the RecoveryAddress can do `let cap = Capability::acquire<KeyRotCap>(recovery_address_account, address_to_recover)`.